adapters/adform/adform.go

```diff
@@ -0,0 +1,364 @@
+	}
+	uri := a.URI
+	if r.isSecure {
+		uri = strings.Replace(uri, "http://", "https://", 1)
+	}
+	return fmt.Sprintf("%s/?CC=1&rp=4&fd=1&stid=%s&%s", uri, r.tid, strings.Join(adUnitsParams, "&"))
+}
+
+func (r *adformRequest) buildAdformHeaders(a *AdformAdapter) http.Header {
+	header := http.Header{}
+
+	header.Set("Content-Type", "application/json;charset=utf-8")
+	header.Set("Accept", "application/json")
+	header.Set("X-Request-Agent", fmt.Sprintf("PrebidAdapter %s", a.version))
+	header.Set("User-Agent", r.userAgent)
+	header.Set("X-Forwarded-For", r.ip)
+	if r.referer != "" {
+		header.Set("Referer", r.referer)
+	}
+	if r.userId != "" {
+		header.Set("Cookie", fmt.Sprintf("uid=%s", r.userId))
+	}
+
+	return header
+}
+
+func parseAdformBids(response []byte) ([]*adformBid, error) {
+	var bids []*adformBid
+	if err := json.Unmarshal(response, &bids); err != nil {
+		return nil, err
+	}
+
+	return bids, nil
+}
+
+// BIDDER Interface
+
+func NewAdformBidder(client *http.Client, endpointURL string) *AdformAdapter {
+	a := &adapters.HTTPAdapter{Client: client}
+
+	return &AdformAdapter{
+		http:    a,
+		URI:     endpointURL,
+		version: "0.1.0",
+	}
+}
+
+func (a *AdformAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	adformRequest, errors := openRtbToAdformRequest(request)
+	if len(adformRequest.adUnits) == 0 {
+		return nil, errors
+	}
+
+	requestData := adapters.RequestData{
+		Method:  "GET",
+		Uri:     adformRequest.buildAdformUrl(a),
+		Body:    nil,
+		Headers: adformRequest.buildAdformHeaders(a),
+	}
+
+	requests := []*adapters.RequestData{&requestData}
+
+	return requests, errors
+}
+
+func openRtbToAdformRequest(request *openrtb.BidRequest) (*adformRequest, []error) {
+	adUnits := make([]*adformAdUnit, 0, len(request.Imp))
+	errors := make([]error, 0, len(request.Imp))
+	for _, imp := range request.Imp {
+		if imp.Banner == nil {
+			errors = append(errors, fmt.Errorf("Adform adapter supports only banner Imps for now. Ignoring Imp ID=%s", imp.ID))
+			continue
+		}
+
+		params, _, _, err := jsonparser.Get(imp.Ext, "bidder")
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		var adformAdUnit adformAdUnit
+		if err := json.Unmarshal(params, &adformAdUnit); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		mid, err := adformAdUnit.MasterTagId.Int64()
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		if mid <= 0 {
+			errors = append(errors, fmt.Errorf("master tag(placement) id is invalid=%s", adformAdUnit.MasterTagId))
+			continue
+		}
+
+		adformAdUnit.bidId = imp.ID
+		adformAdUnit.adUnitCode = imp.ID
+		adUnits = append(adUnits, &adformAdUnit)
+	}
+
+	referer := ""
+	if request.Site != nil {
+		referer = request.Site.Page
+	}
+
+	tid := ""
+	if request.Source != nil {
+		tid = request.Source.TID
+	}
+
+	secure := false
+	for _, imp := range request.Imp {
```

Comment:
just a suggestion, maybe you could check if it's secure in the for loop above (line 263) since you're already looping through the imps.

Reply:
updated